Story Cards #15 and #17

KamilRosiak opened the pull request:
Story Card #15: Show the constraint explanation icons in the new view similarly to how it was implemented in constraints beneath the model
Story Card #17: When selecting a constraint in the new view the corresponding features are highlighted in the model.

Commit:
new Development Branch

Commit:
- Refactoring of the IPartListner into a own class
ConstraintViewPartListener

Commit:
finished function that highlights features with a border when the
related constraint is selected

Commit:
refactored code

Commit:
fixed a NullPointerException

Commit:
Story card 15: -Constructed Pair class
- Added icons with confidence color

Commit:
Highlighting Features merged on developmentBranch

Commit:
- added redundant feature to explanation.

Commit:
Merge branch 'DevelopmentBranch' of https://github.com/oliviia12/FeatureIDE.git into DevelopmentBranch

Commit:
ID change of editconstraintinviewaction

Commit:
- removed a code clone

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/ConstraintViewController.java:
```diff
@@ -259,14 +212,47 @@ private void checkForRefresh() {
+	}
+
+	/**
+	 * returns a set of constraints if a explanation is available else null
```
a**n** explanation

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/ConstraintViewController.java:
```diff
@@ -259,14 +212,47 @@ private void checkForRefresh() {
+				final Explanation explanation = (Explanation) fme.diagramEditor.getActiveExplanation();
+				viewer.removeAll();
+				final List<ConstraintColorPair<IConstraint, Color>> constraintList = new ArrayList<ConstraintColorPair<IConstraint, Color>>();
+				for (final Object rea : explanation.getReasons()) {
```
More descriptive names (Object rea)

ThisIsATunaSandwich replied:
Maybe change fmr and fme to fMR and fME for better readability

PaulWestphal replied:
I think lower case is fine for three letter long abbreviations.

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/ConstraintViewController.java:
```diff
@@ -284,6 +270,28 @@ public void pageChanged(PageChangedEvent event) {
+				final IConstraint constraint = (IConstraint) treeSelection.getFirstElement();
+				for (final IFeature feature : FeatureModelUtil.getFeatureModel().getFeatures()) {
+					graphfeature = FeatureModelUtil.getActiveFMEditor().diagramEditor.getGraphicalFeatureModel().getGraphicalFeature(feature);
+					graphmodel = FeatureModelUtil.getActiveFMEditor().diagramEditor.getGraphicalFeatureModel();
```
graphfeature and graphmodel into graphFeature and graphModel

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/listener/ConstraintViewPartListener.java:
```diff
@@ -0,0 +1,105 @@
+import de.ovgu.featureide.fm.ui.views.constraintview.ConstraintViewController;
+
+/**
+ * TODO description
```
Add class description

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintColorPair.java:
```diff
@@ -0,0 +1,52 @@
+package de.ovgu.featureide.fm.ui.views.constraintview.util;
+
+/**
+ * TODO description
```
add class description

ThisIsATunaSandwich commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java:
```diff
@@ -70,20 +73,54 @@ public ConstraintView(Composite parent) {
+			elementImg = FM_INFO;
+		} else {
+			elementImg = new Image(Display.getDefault(), IMAGE_EMPTY.getImageData());
+			final GC gc = new GC(elementImg);
```
more descriptive name

ThisIsATunaSandwich reviewed (commented):
Code looks good and works.
Some names could be more descriptive. 
Furthermore, class descriptions should be added.

@PaulWestphal

PaulWestphal commented:
I just noticed that the constraint view is not reset when you delete the feature model. This also causes a nullpointerexception when you click on one of the features.

PaulWestphal commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintColorPair.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @author "Rosiak Kamil"
+ */
+public class ConstraintColorPair<S, T> {
```
Eclipse has pairs (org.eclipse.debug.internal.ui.Pair), wouldn't that work? I think you're reinventing the wheel here.

PaulWestphal commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java:
```diff
@@ -70,20 +73,54 @@ public ConstraintView(Composite parent) {
+			elementImg = new Image(Display.getDefault(), IMAGE_EMPTY.getImageData());
+			final GC gc = new GC(elementImg);
+			gc.setBackground(color);
+			// MAGIC NUMBERS
```
You can remove this now.

PaulWestphal commented:
We have to discuss your Pair class tomorrow.

tthuem commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/ConstraintViewController.java:
```diff
@@ -259,14 +212,47 @@ private void checkForRefresh() {
+	 * returns a set of constraints if a explanation is available else null
+	 *
+	 */
+	@SuppressWarnings("rawtypes")
```
Would be better to avoid that by means of explicit casts.

tthuem commented on plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/featuremodeleditview/ExampleParent.java:
```diff
@@ -78,8 +78,6 @@ public void initChildren() {
 				example = c.calculateExample(added);
-			}
-			if (example == null) {
 				addChild(NONE);
```
Be careful! This change is not semantics preserving! Please revert those changes unless you have clear evidence that it improves the behavior.

Commit:
- added all changes from git.
- refactoring of the Pair class so that all values get a declarative
name.
- revert the change in the ExampleParent.

Commit:
- added a javadoc comment

Commit:
- added javadoc comment
